Cache secret lookups, fixes #2924

ralekseenkov opened the pull request:
refactoring of secret managers + generic secret cache implementaiton

this will also serve as a good foundation for team-based credential managers RFC, as now you can introduce a composable secret manager (same way as caching and retrying is implemented), which can iterate over several underlying secret managers

```
* composable secret manager for team 1
           -> vault 1
           -> vault 2
           -> ssm
```                 

and you can add retries and caching at any layer. e.g. one global cache, or several caches per each credential manager, etc.

see commit description

topherbullock commented:
Hey @ralekseenkov 
Question about one of the commits in this PR: https://github.com/concourse/concourse/pull/3628/commits/c0cecccb6cffe6e75cfd530f798aeb2fe8d42dbf - was there a need to add the `gobin ...` directives vs installing `counterfeiter` to the `GOPATH/$bin`(https://github.com/maxbrunsfeld/counterfeiter#installing-counterfeiter-to-gopathbin), or is this change to ensure the version of the tool is tracked `go.mod`?

ralekseenkov commented:
@topherbullock my PR was created before Concourse team opted to install counterfeiter to $GOPATH/bin. I just rebased it on top of master and removed these changes

vito commented:
Taking a look at this now! There are merge conflicts but I've resolved them locally (it wasn't too bad). If all is good I'll just push the merge with the conflict resolutions. :+1: 

So far I've just kicked the tires at a surface level and confirmed that credentials are cached with a locally running Vault by inspecting the audit log to see when credentials are fetched.

The default cache duration is only one minute, which means credentials will still be re-fetched roughly every resource check (which also has a one minute interval). To me that makes sense, from the standpoint that it's a conservative default that will help when there are many builds running, while still preventing a credential rotated independent of its lease duration (i.e. manually) from impacting the system for too long, as the next resource check or two will pick it up. If users need to mitigate load they can always increase the `--secret-cache-duration`.

I haven't done any testing on lease durations yet, but its interaction with `--secret-cache-duration` makes sense to me (smaller duration wins).

Thanks for starting on all this, and thanks for your patience!

vito commented on atc/atccmd/command.go:
```diff
@@ -880,14 +880,20 @@ func (cmd *RunCommand) variablesFactory(logger lager.Logger) (creds.VariablesFac
+	result := secretsFactory.NewSecrets()
+	result = creds.NewRetryableSecrets(result, cmd.CredentialManagement.RetryConfig)
+	if cmd.CredentialManagement.CacheConfig.Enabled {
+		result = creds.NewCachedSecrets(result, cmd.CredentialManagement.CacheConfig)
```
:tada: composition!

vito commented on atc/creds/cached_secrets.go:
```diff
@@ -0,0 +1,73 @@
+package creds
+
+import (
+	"github.com/patrickmn/go-cache"
```
nice that there was a package for this :+1:

vito commented on atc/creds/cached_secrets.go:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	// here we want to cache secret value, expiration, and found flag too
+	// meaning that "secret not found" responses will be cached too!
```
neat

vito commented on atc/creds/secrets_factory.go:
```diff
@@ -0,0 +1,25 @@
+	Get(string) (interface{}, *time.Time, bool, error)
+
+	// NewSecretLookupPaths returns an instance of lookup policy, which can transform pipeline ((var)) into one or more secret paths, based on team name and pipeline name
+	NewSecretLookupPaths(string, string) []SecretLookupPath
```
:+1: nice idea - simplifies everyone's `Get` implementation

although, should this just be called `SecretLookupPaths`? it doesn't really feel like a constructor to me

vito commented on atc/creds/credhub/credhub.go:
```diff
@@ -87,14 +85,3 @@ func (c CredHubAtc) findCred(path string) (credentials.Credential, bool, error)
-	return path.Join(append([]string{c.PathPrefix}, segments...)...)
-}
-
-func (c CredHubAtc) List() ([]template.VariableDefinition, error) {
```
:+1: I like that folks don't have to implement this stub anymore

vito commented on atc/creds/secretsmanager/secretsmanager.go:
```diff
@@ -3,69 +3,60 @@ package secretsmanager
+func (s *SecretsManager) Get(secretPath string) (interface{}, *time.Time, bool, error) {
+	value, expiration, found, err := s.getSecretById(secretPath)
+	if err != nil {
+		s.log.Error("unable to retrieve aws secret", err, lager.Data{
```
this should be conventional formatting:

```suggestion
		s.log.Error("failed-to-fetch-aws-secret", err, lager.Data{
```

ralekseenkov commented:
Sounds good. I was trying to be careful, decoupling variable lookup logic (i.e. NewSecretLookupPaths) from secret retrieval. Hopefully nothing got broken as a result of this.

But there are cases like Kubernetes credential manager - it doesn't seem to have any tests and it's currently missing docs, see https://github.com/concourse/concourse/issues/2003. That makes it challenging to verify

ralekseenkov commented:
thanks :)

samgurtman commented:
I hate to do this, but would anyone have any guidance on back porting this to 5.1.0? I have a big of egg on my face after moving our credentials to SSM. I guess I can set the default resource check to much longer, but this seems like a way better solution.

ralekseenkov commented:
@samgurtman it might be easier to wait for 5.2.0 because maintaining your own forks is never fun. if you really want to backport, ping me on Discord and I can try to help

cirocosta commented:
Hey @ralekseenkov, I noticed that the kubernetes cred mgmt changed a bit, breaking the tests we have in place for such integration:

https://github.com/concourse/concourse/blob/37dd02b20741a9c90b362f3f4edeb918a674caa9/topgun/k8s/kubernetes_creds_mgmt_test.go#L73-L120

Seems related to the new idea of having the `:` separator here:

https://github.com/concourse/concourse/blob/7b846eeac9a5ca232d521f25b2c0c410e72e02de/atc/creds/kubernetes/kubernetes.go#L33-L35

Was there a specific reason for going this way?

Thanks!

cirocosta commented:
Oh, here's an example failure:

```
 running: kubectl create secret generic foo --namespace=topgun-k8s-cm-664743-1-main --from-literal=value=bar
secret/foo created
running: kubectl create secret generic caz --namespace=topgun-k8s-cm-664743-1-main --from-literal=baz=zaz
secret/caz created

fly -t concourse-topgun-k8s-1 set-pipeline -n -c ../pipelines/minimal-credential-management.yml -p pipeline
jobs:
  job unit has been added:
+ name: unit
+ plan:
+ - task: run
+   config:
+     platform: linux
+     image_resource:
+       type: registry-image
+       source:
+         repository: busybox
+     run:
+       path: /bin/sh
+       args:
+       - -c
+       - |
+         echo plain-secret=((foo))
+         echo nested-secret=((caz.baz))

...

failed to interpolate task config: 
  Finding variable 'foo': unable to split kubernetes secret path 
  into [namespace]:[secret]: topgun-k8s-cm-664743-1-/main/pipeline/foo
```

ralekseenkov commented:
@cirocosta it should be backwards compatible. I'll look into this today
